### DISCUSSION TITLE
Publishing extension failure

### DISCUSSION THREAD

**Discussion by will-jai** [iTwin/itwinjs-core#1698] [Q&A]
Hi,
We got a question regarding iTwin viewer, our application allows users to provide clientid, imodelid, and contextid at the run time of our application and we display the corresponding iModel at runtime, for this approach we are trying to make the configuration as generic as possible. This means we cannot pre-write the code which would interact with the model or display data on it, as we don't even know what model user will try to embed. 

 

We are hoping there would be a way where we can just pass the data either to the viewer or the iTwin server and the model would already know how to render or display it. Please advise if there is a way which can allow us to achieve this?

Thanks,
Will

**Comment by aruniverse** [comment]
Sorry, not sure if I understood the question correctly, but the `Viewer` component exposes a prop for the context and iModel ids, so couldnt you pass them like you want? It should just open the default view of that model. And if you wanted to override it, you theoretically could because there is a prop we expose for that, but i'm guessing that won't be generic enough.

**Reply by will-jai** [reply]
Thanks for the reply, we have done the part of displaying IModel, and let me clarify the question, if we want to display live data from a data source, is that has to write an extension to display the information? or there is a way we can directly send the data to the viewer / the imodel hub and the iModel would already know how to render it in the viewer? if an extension is required, is there a way we can allow the user to upload a JavaScript file and register the JavaScirpt in the viewer at runtime?

**Reply by aruniverse** [reply]
So you wouldn't be able to send external data from a live data source to the imodel hub.
I would recommend using an extension or UiItemsProvider to add to your viewer, which can read live data from your external source. I believe we have a good example of this use case in this [Wind Farm IoT sample](https://github.com/iModeljs-meets-AzureDT/windfarm-iot). I would also recommend going through the iTwin Accreditation program which walks you through a good example of adding data to your viewer.

We don't provide a way for users to upload a file for an extension to be registered at runtime, but it's probably doable on your end. We do provide api's to fetch an extension at runtime, I recommend look at the [Extension apis](https://www.itwinjs.org/reference/imodeljs-frontend/extensions/extensionadmin/).

**Reply by will-jai** [reply]
Hi, 

I was trying to publish an extension via extension-cli, however, i got the below error messages. Could you please advise how I can resolve this? thanks

Without Context Id:
     Unknown error: Could not find a team for the current user, 
With Context Id 
o	Error: Unknown error from backend: Server returned status: 403, message: User is not allowed to modify using Asset/Project context

Upload to public context
o	Error: Unknown error from backend: Server returned status: 403, message: User is not allowed to modify public context

**Reply by aruniverse** [reply]
@paulius-valiunas ?

@will-jai have you looked at this discussion post? https://github.com/imodeljs/imodeljs/discussions/1613#discussioncomment-871995

**Reply by paulius-valiunas** [reply]
@calebmshafer any idea why a user wouldn't have a team id assigned?

**Reply by will-jai** [reply]
@aruniverse, I've looked at the discussion post, unfortunately, I got the above errors when I tried to publish an extension.

**Reply by calebmshafer** [reply]
The first error of not finding a Team is interesting and would need some more information to track down. @will-jai lets follow up on this part in our call next week.

The rest of the error I believe is a permissions issue. In order to publish an Extension for your Organization you must have the "Manage Extension" Enterprise-level role assigned to you. That role will need to be set by an administrator of your CONNECT Org and is under the "Extension Service" heading.  We can follow up on this as well when we cover the first issue.
